### ISSUE TITLE
Errors thrown by papiea-client should be well-defined

### ISSUE THREAD

**Issue by joshua-berry-ntnx** [nutanix/papiea-js#376]
[This might be a dup of or related to #255.]

tl;dr: I would like papiea-client to catch and re-throw (or re-reject) axios failures with some types exposed through the papiea-client interface which are specific to the error that occurred.

Right now it looks like papiea-client simply allows axios exceptions to bubble up to the caller. The problem with this is the caller is probably in another module which depends on papiea-client, but may not depend on axios. So it's very difficult for callers to name the type of exception they are catching, e.g.:

```ts
try {
    // ... papiea-client call ...
} catch (e) {
    if (e instanceof ???) {
        ...
    } else { throw e; }
}
```

Similarly, there are several different kinds of errors thrown by papiea-client (and/or papiea-engine) in different situations. For example, update conflicts, poorly-formatted requests, authorization errors, etc. These errors should probably be distinguishable as Papiea errors rather than generic HTTP errors, network errors, etc., and it is often useful to distinguish between different kinds of Papiea errors (poorly-formatted spec, update conflict, etc.).

Some ideas/examples worth considering (I'm mainly focusing on `update()` because that's where this first occurred to me), feel free to ignore my suggested names:

- `InvalidSpecError` - entity spec validation failed in update() or create(), perhaps with additional details about what failed in the error message
- `UpdateConflictError` - update() with old/incorrect spec_version()
- `NotFoundError` - get() which returns no entity (or maybe this can just return undefined, if that's the current behavior)
- `NotAuthorizedError`

**Comment by joshua-berry-ntnx** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Address Papiea issues needed for inventory-only BMaaS](https://trello.com/c/V6gmTU73/66-address-papiea-issues-needed-for-inventory-only-bmaas)

**Comment by Emulebest** [comment]
Yes, this is totally related to #255
